homeassistant/components/ecobee/switch.py

```diff
@@ -0,0 +1,105 @@
+"""Support for using switch with ecobee thermostats."""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from dataclasses import dataclass
+from datetime import datetime
+import logging
+from typing import Any
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import EcobeeData
+from .const import DOMAIN
+from .entity import EcobeeBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+DATE_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+
+@dataclass(frozen=True, kw_only=True)
+class EcobeeSwitchEntityDescription(SwitchEntity):
+    """Class describing Ecobee switch entities."""
+
+    ecobee_setting_key: str
+    set_fn: Callable[[EcobeeData, int, int], Awaitable]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the ecobee thermostat switch entity."""
+    data: EcobeeData = hass.data[DOMAIN]
+    _LOGGER.debug("Adding ventilators 20 min switch if present")
+
+    async_add_entities(
+        (
+            EcobeeVentilator20MinSwitch(data, index)
+            for index, thermostat in enumerate(data.ecobee.thermostats)
+            if thermostat["settings"]["ventilatorType"] != "none"
+        ),
+        True,
+    )
+
+
+class EcobeeVentilator20MinSwitch(EcobeeBaseEntity, SwitchEntity):
+    """A Switch class, representing 20 min timer for an ecobee thermostat with ventilator attached."""
+
+    def __init__(
+        self,
+        data: EcobeeData,
+        thermostat_index: int,
+    ) -> None:
+        """Initialize ecobee ventilator platform."""
+        super().__init__(data, thermostat_index)
+        self._attr_unique_id = "ventilator_20m_timer"
+        self._attr_native_value = False
+        self.update_without_throttle = False
+        self._time_zone_delay = datetime.strptime(
+            self.thermostat["utcTime"], DATE_FORMAT
+        ) - datetime.strptime(self.thermostat["thermostatTime"], DATE_FORMAT)
```

Comment:
Caching the time zone delay will cause problems for a system that runs across a daylight savings time boundary.  We should calculate this when needed.

Reply:
Not sure I follow you. The way I see this it will match whatever time the Ecobee is operating on. Since all the timestamps comes from the device

Reply:
It looks like you are calculating the time zone delay from UTC.  This delay changes when daylight savings time engages / disengages.  In New York time the offset right now is 4 hours.  But on 11/4 that becomes 5 hours. If I start HA on 11/2 it'll record 4 hours in the __init__ for the entity.  Two days later that calculation would yield 5 hours but 4 is cached.  As a result the switch will reporting will be off until HA is restarted.

Reply:
ooo, I see your point, I can move that calculation in every update. What do you think. 

I assumed it would be a lot of unnecessary recalculation. But you are right, this would get broken twice a year. 


Do you think it's worth moving in the `async_update`

Reply:
Based on our discussion I have proceed to move that logic in the update function